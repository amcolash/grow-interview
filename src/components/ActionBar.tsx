import React from 'react';
import { style } from 'typestyle';

const barStyle = style({
  display: 'flex',
  width: '100%',
  height: '96px',
  padding: '16px',
  justifyContent: 'center',
  alignItems: 'center',
  gap: '16px',

  borderRadius: '100px',
  background: 'var(--neutral-000, #FFF)',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function ActionBar() {
  return <div className={barStyle}></div>;
}
