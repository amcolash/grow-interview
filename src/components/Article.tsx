import React from 'react';
import { style } from 'typestyle';

import { ArticleData } from '../hooks/useFetchArticles';

interface ArticleProps {
  article: ArticleData;
}

const articleStyle = style({
  display: 'flex',
  padding: '24px',
  alignItems: 'center',
  gap: '20px',
  alignSelf: 'stretch',
  fontSize: '16px',
  lineHeight: '24px',

  borderRadius: '12px',
  border: '1px solid var(--neutral-300, #E7EAEB)',
  background: 'var(--white, #FFF)',

  $nest: {
    '& .index, & .views': {
      color: 'var(--neutral-500, #A7AAAB)',
    },
    '.title': {
      flex: 1,
    },
    '.views': {
      fontSize: '14px',
      fontFamily: 'Poppins',
      lineHeight: '16px',
    },
  },
});

export function Article(props: ArticleProps) {
  const title = props.article.article.replace(/_/g, ' ');

  return (
    <article className={articleStyle} data-testid="article">
      <div className="index">{props.article.rank}</div>
      <div className="title">{title}</div>
      <div className="views">{props.article.views.toLocaleString('en-US')} views</div>
    </article>
  );
}
