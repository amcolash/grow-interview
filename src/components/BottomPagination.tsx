import React from 'react';
import ReactPaginate from 'react-paginate';
import { style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface BottomPaginationProps {
  totalPages: number;
  currentPage: number;
  onPageChange: (selectedItem: { selected: number }) => void;
}

const pageListStyle = style({
  display: 'flex',
  justifyContent: 'center',
  flexWrap: 'wrap',
  gap: '8px',
  listStyle: 'none',
  margin: 0,
  padding: 0,

  $nest: {
    '& a': {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '40px',
      height: '40px',
      padding: '10px',
      color: 'var(--neutral-900, #05090D)',
      fontFamily: 'Poppins',
    },
    '& li': {
      border: '1px solid var(--neutral-400, #D4D8D9)',
      background: 'var(--neutral-000, #FFF)',
      borderRadius: '100px',
      userSelect: 'none',

      $nest: {
        '&.disabled': {
          background: 'var(--neutral-400, #D4D8D9)',
        },
        '&.disabled > a': {
          color: 'var(--neutral-600, #737680)',
        },
        '&:not(.disabled,.selected):hover': {
          background: 'var(--neutral-300, #e7eaeb)',
        },
        '&.selected': {
          background: 'var(--avocado-300, #E0E9CB)',
        },
        '&.selected > a': {
          color: 'var(--brand-green-500, #025B4B)',
        },
      },
    },
  },
});

export function BottomPagination(props: BottomPaginationProps) {
  return (
    <ReactPaginate
      className={pageListStyle}
      breakLabel="..."
      nextLabel={<ChevronRight />}
      onPageChange={props.onPageChange}
      pageCount={props.totalPages}
      previousLabel={<ChevronLeft />}
      renderOnZeroPageCount={null}
    />
  );
}
