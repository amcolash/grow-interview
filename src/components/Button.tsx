import React, { ButtonHTMLAttributes } from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const buttonStyle = style(
  {
    display: 'flex',
    width: '160px',
    padding: '12px 24px',
    justifyContent: 'center',
    alignItems: 'center',
    gap: '8px',
    flexShrink: '0',
    alignSelf: 'stretch',
    borderRadius: '100px',
    background: 'var(--brand-green-500, #025b4b)',
    border: 'none',
    color: 'var(--neutral-000, #fff)',
    textAlign: 'center',
    fontSize: '16px',
    fontFamily: 'Poppins',
    fontWeight: '500',
    lineHeight: '24px',
    letterSpacing: '-0.32px',
  },
  media({ maxWidth: mobileWidth }, { width: '100%' })
);

export function Button(props: ButtonHTMLAttributes<HTMLButtonElement> & { 'data-testid'?: string }) {
  return <button {...props} className={buttonStyle} data-testid={props['data-testid'] || 'button'} />;
}
