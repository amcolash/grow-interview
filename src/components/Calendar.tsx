import React from 'react';
import { CaptionProps, DayPicker, useNavigation } from 'react-day-picker';
import 'react-day-picker/dist/style.css';
import { cssRule, style } from 'typestyle';

import { ReactComponent as ChevronLeft } from '../icons/chevron-left.svg';
import { ReactComponent as ChevronRight } from '../icons/chevron-right.svg';

interface CalendarProps {
  date: Date;
  onChange: (date: Date) => void;
}

cssRule('.rdp', {
  fontFamily: 'Poppins',
  margin: 0,
});

export function Calendar(props: CalendarProps) {
  return (
    <DayPicker
      mode="single"
      selected={props.date}
      onDayClick={(day) => props.onChange(day)}
      disabled={{ from: new Date(), to: new Date(9999, 12, 31) }}
      formatters={{
        formatWeekdayName: (weekday) =>
          weekday.toLocaleDateString(undefined, {
            weekday: 'short',
          }),
      }}
      styles={{
        head_cell: {
          fontWeight: 'normal',
          color: 'var(--neutral-500, #A7AAAB)',
          fontSize: '12px',
          lineHeight: '18px',
        },
        cell: {
          fontSize: '14px',
          lineHeight: '22px',
        },
      }}
      components={{
        Caption: CustomCaption,
      }}
      modifiersStyles={{
        selected: {
          color: 'var(--brand-green-500, #025B4B)',
          background: 'var(--ivy-300, #B3CEC9)',
          fontWeight: '500',
        },
      }}
    />
  );
}

const captionStyle = style({
  display: 'flex',
  justifyContent: 'space-between',
  marginBottom: '32px',

  $nest: {
    '& button': {
      border: 'none',
      background: 'none',
      display: 'flex',
      padding: '0',
      alignItems: 'center',
      width: '24px',
      height: '24px',
      justifyContent: 'center',
      borderRadius: '100%',
    },
    '& button:hover': {
      background: 'var(--neutral-300, #e7eaeb)',
    },
    '& label': {
      fontSize: '16px',
      lineHeight: '24px',
      fontWeight: '500',
    },
  },
});

function CustomCaption(props: CaptionProps) {
  const { previousMonth, nextMonth, goToMonth, displayMonths } = useNavigation();
  const caption = props.displayMonth.toLocaleDateString(undefined, {
    month: 'long',
    year: 'numeric',
  });

  return (
    <div className={captionStyle}>
      <button
        onClick={() => {
          if (!previousMonth) return;
          goToMonth(previousMonth);
        }}
      >
        <ChevronLeft />
      </button>
      <label>{caption}</label>
      <button
        onClick={() => {
          if (!nextMonth) return;
          goToMonth(nextMonth);
        }}
        disabled={nextMonth ? nextMonth.getTime() > Date.now() : false}
      >
        <ChevronRight />
      </button>
    </div>
  );
}
