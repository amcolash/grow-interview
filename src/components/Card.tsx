import React from 'react';
import { style } from 'typestyle';

interface CardProps {
  children: React.ReactNode;
}

const cardListStyle = style({
  display: 'flex',
  flexDirection: 'column',
  gap: '20px',
  padding: '32px',

  borderRadius: '16px',
  background: '#FFF',
  boxShadow: '0px 2px 0px 1px rgba(5, 9, 13, 0.06)',
});

export function Card(props: CardProps) {
  return <div className={cardListStyle}>{props.children}</div>;
}
