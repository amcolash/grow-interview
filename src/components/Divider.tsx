import React from 'react';
import { media, style } from 'typestyle';

import { mobileWidth } from '../consts';

const dividerStyle = style(
  {
    borderLeft: '1px solid var(--neutral-300, #E7EAEB)',
    height: '64px',
    margin: '0 16px',
  },
  media({ maxWidth: mobileWidth }, { display: 'none' })
);

export function Divider() {
  return <div className={dividerStyle}></div>;
}
