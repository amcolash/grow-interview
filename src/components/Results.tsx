import React from 'react';

import { ArticleData } from '../hooks/useFetchArticles';
import { Article } from './Article';
import { Card } from './Card';

interface ResultsProps {
  url?: string;
  error?: string;
  currentPage?: ArticleData[];
}

export function Results(props: ResultsProps) {
  const { currentPage, error, url } = props;

  if (!url || url.length === 0) return null;
  if (error || !currentPage)
    return (
      <Card testId="search-results">
        <div style={{ textAlign: 'center' }} data-testid="results-error">
          Something went wrong fetching articles!
          <pre>{error}</pre>
        </div>
      </Card>
    );

  return (
    <Card testId="search-results">
      {currentPage.map((article) => (
        <Article key={article.rank} article={article} />
      ))}
    </Card>
  );
}
