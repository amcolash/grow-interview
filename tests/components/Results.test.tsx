import { render } from '@testing-library/react';
import React from 'react';

import { Results } from '../../src/components/Results';

describe('Results', () => {
  it('renders null by default', () => {
    const { asFragment } = render(<Results />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders errors', () => {
    const { asFragment } = render(<Results url="http://testurl.com" error="Something went wrong getting data" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders error (no page)', () => {
    const { asFragment } = render(<Results url="http://testurl.com" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders valid results', () => {
    const { asFragment } = render(
      <Results
        url="http://testurl.com"
        currentPage={[
          { article: 'Test1', rank: 1, views: 123456 },
          { article: 'Test2', rank: 2, views: 123 },
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
