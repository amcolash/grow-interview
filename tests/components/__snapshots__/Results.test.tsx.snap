// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Results > renders error (no page) 1`] = `
<DocumentFragment>
  <div
    class="fzx2n3q"
    data-testid="search-results"
  >
    <div
      data-testid="results-error"
      style="text-align: center;"
    >
      Something went wrong fetching articles!
      <pre />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Results > renders errors 1`] = `
<DocumentFragment>
  <div
    class="fzx2n3q"
    data-testid="search-results"
  >
    <div
      data-testid="results-error"
      style="text-align: center;"
    >
      Something went wrong fetching articles!
      <pre>
        Something went wrong getting data
      </pre>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Results > renders null by default 1`] = `<DocumentFragment />`;

exports[`Results > renders valid results 1`] = `
<DocumentFragment>
  <div
    class="fzx2n3q"
    data-testid="search-results"
  >
    <article
      class="f1cqgvxd"
      data-testid="article"
    >
      <div
        class="index"
      >
        1
      </div>
      <div
        class="title"
      >
        Test1
      </div>
      <div
        class="views"
      >
        123,456 views
      </div>
    </article>
    <article
      class="f1cqgvxd"
      data-testid="article"
    >
      <div
        class="index"
      >
        2
      </div>
      <div
        class="title"
      >
        Test2
      </div>
      <div
        class="views"
      >
        123 views
      </div>
    </article>
  </div>
</DocumentFragment>
`;
